### PR TITLE
fp16 support for FullyConnected op

### PR DESCRIPTION
**Pull request by slayton58**
Includes math lib support, removal of double-precision.

### CODE REVIEW

File: caffe2/operators/utility_ops_gpu.cc [hunk 1]
```diff
@@ -5,6 +5,28 @@
...
+
+template <>
+bool SumOp<CUDAContext>::RunOnDevice() {
+  if (Input(0).IsType<float>()) {
+    return DoRunWithType<float,float>();
+  } else if (Input(0).IsType<float16>()) {
+    return DoRunWithType<float16,float16>();
+  } else {
+    CAFFE_THROW("Unsupported inputs");
+  }
```
Comment (asaadaldien) [inline caffe2/operators/utility_ops_gpu.cc:27]: return false;

**Review by asaadaldien — COMMENTED** [review]
Thanks @slayton58 for working on this. There is some conflicts. Can you please rebase?

**Comment by slayton58** [comment]
rebased + return false changes made

**Review by asaadaldien — COMMENTED** [review]
LGTM! Thanks @slayton58 . few nits. can you also rebase before I import this :).

File: caffe2/utils/conversions.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#else
+#define CONVERSIONS_DECL inline
+#endif
+
+namespace caffe2 {
+
+namespace convert {
+
+namespace {
+inline float16 cpu_float2half_rn(float f)
```
[the remark this answers is not in the document]
Comment (slayton58) [reply caffe2/utils/conversions.h]: Fixed just now - I rebased just before your comment, so I hope that's still good :)

File: caffe2/operators/elementwise_op.cu [hunk 3]
```diff
@@ -222,6 +220,56 @@ bool SumReduceLikeOp<CUDAContext>::DoRunWithType() {
...
 }
 
+namespace {
+
 REGISTER_CUDA_OPERATOR(SumReduceLike, SumReduceLikeOp<CUDAContext>);
 
+template <typename T, typename M>
+__global__
+void binary_add_kernel(const int N, const T* a, const T* b, T* r) {
+  for (int idx=threadIdx.x+blockIdx.x*blockDim.x; idx < N; idx+=blockDim.x*gridDim.x) {
```
[the remark this answers is not in the document]
Comment (slayton58) [reply caffe2/operators/elementwise_op.cu]: Missed this before - fixed

**Review by asaadaldien — COMMENTED** [review]
@slayton58 
Can you please update the signature for Gemm & Gemv
Here:
caffe2/caffe2/utils/math_cpu.cc:261
caffe2/caffe2/utils/math_cpu.cc:292
caffe2/caffe2/contrib/nervana/nervana_math_gpu.cc:13 

Also update the call to FullyConnectedOp here
caffe2/caffe2/contrib/nervana/nervana_fc_op_gpu.cc:8

File: caffe2/operators/fully_connected_op_gpu.cc [hunk 4]
```diff
@@ -2,9 +2,51 @@
...
+                         float,    // Y
+                         float>(); // Math
+  } else if (Input(0).IsType<float16>()) {
+    return DoRunWithType<float16,    // X
+                         float16,    // W
+                         float16,    // B
+                         float16,    // Y
+                         float>(); // Math
+  }
+}
```
[the remark this answers is not in the document]
Comment (asaadaldien) [reply caffe2/operators/fully_connected_op_gpu.cc]: else {
CAFFE_THROW("Unsupported type");
return false;
}

[the remark this answers is not in the document]
Comment (slayton58) [reply caffe2/operators/fully_connected_op_gpu.cc]: I assume the nervana change was meant to be caffe2/contrib/nervana/nervana_math_gpu.cc

Changes made.

File: caffe2/utils/conversions.h [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+namespace caffe2 {
+
+namespace convert {
+
+namespace {
+inline float16 cpu_float2half_rn(float f) {
+    float16 ret;
+
+    unsigned x = *((int*)(void*)(&f));
```
Comment (Yangqing) [inline caffe2/utils/conversions.h]: Just for safety... could you do

    static_assert(sizeof(unsigned int) == sizeof(float),
        "Programming error: sizeof(unsigned int) != sizeof(float).");
    unsigned x = *(reinterpret_cast<unsigned int*>(&f));

Comment (asaadaldien) [reply caffe2/utils/conversions.h]: Unfortunately both `*((int*)(void*)(&f))` and `*(reinterpret_cast<unsigned int*>(&f))` will break builds with -fstrict-aliasing (error: dereferencing type-punned pointer will break strict-aliasing rules)
To fix that we can just use the pointer itself: 
`unsigned int* x = reinterpret_cast<unsigned int*>(&f);`
 and in line 101 as well.

Comment (slayton58) [reply caffe2/utils/conversions.h]: I don't understand how this is meant to work if I can't dereference that pointer -- I need the actual value of x as originally done.

Comment (asaadaldien) [reply caffe2/utils/conversions.h]: @slayton58 You can dereference the pointer x later but the compiler with -fstrict-aliasing
is not happy with this `*(reinterpret_cast<unsigned int*>(&f))`

Comment (slayton58) [reply caffe2/utils/conversions.h]: ok, so I can do `unsigned *xp = reinterpret_cast<unsigned int*>(&f); unsigned x = *xp;`

**Comment by slayton58** [comment]
Updated: Note that I've reverted the dgemv commit, if we need to support double precision (I was under the impression we no longer did) then there's more work to be done in terms of function signatures.

File: caffe2/utils/conversions.h (hunk 5 above)
Comment (asaadaldien) [reply caffe2/utils/conversions.h]: Yes that should work. You can do this or dereference x in line 33.

**Comment by dragonxlwang** [comment]
Hey, am wondering what's the reason behind the decision to remove double-precision?

**Comment by slayton58** [comment]
@dragonxlwang It complicates the API and no-one uses double precision.

**Pull request closed without merging**